### DISCUSSION TITLE
NER and Relation Extraction pipeline: Could not determine any instances in doc.

### DISCUSSION THREAD

**Discussion by MightyGoldenOctopus** [explosion/spaCy#8499] [Help: Coding & Implementations] [answered]
Hello SpaCy community,

I, a freshly converted SpaCy newbie, am currently trying to plug a pretrained NER model into the relation extraction pipeline, I think I have implemented all the changes recommended in the similar posts on this forum, namely the change in the `score_relations` function to take in account the predictions not in the gold as false positives:
```python
def score_relations(examples: Iterable[Example], threshold: float) -> Dict[str, Any]:
        ...
        for key, pred_dict in pred.items():
            if key not in gold.keys():
                for k, v in pred_dict.items():
                    if v >= threshold:
                        micro_prf.fp += 1
                continue
            gold_labels = [k for (k, v) in gold[key].items() if v == 1.0]
       ...
```
And during training phase i get this:
![rel_training_issue](https://user-images.githubusercontent.com/25876592/123407572-65909f00-d5ac-11eb-90f9-2ad939fd50be.png)

And of course the loss is 0 because the `get_instances` function fails to identify any viable instances, further more all the `doc.ents` processed by this function are without entity and thus empty tuples. I generated the dataset using my own version of `parse_data.py` and I confirmed the samples contained entities and positive relations:
```python
def convert_rel(json_loc: Path, output_path: Path, rel_labels: str):
    REL_LABELS = [label.strip() for label in rel_labels.split(",")]
    Doc.set_extension("rel", default={})
    vocab = Vocab()

    docs = []
    count_sample, count_pos, count_all = 0, 0, 0

    for example in srsly.read_jsonl(json_loc):
        if example["answer"] != "accept":
            continue
        span_starts = set()
        pos, neg = 0, 0
        try:
            # Parse tokens
            words = [t["text"] for t in example["tokens"]]
            spaces = [t["ws"] for t in example["tokens"]]
            doc = Doc(vocab, words=words, spaces=spaces)

            # Parse entities
            spans = example["spans"]
            ents = []
            span_end_to_start = {}
            for span in spans:
                start, end, label = span["start"], span["end"], span["label"]
                entity = doc.char_span(start, end, label=label)
                span_end_to_start[span["token_end"]] = span["token_start"]
                ents.append(entity)
                span_starts.add(span["token_start"])
            doc.set_ents(ents)

            # Parse relations
            rels = {}
            for x1 in span_starts:
                for x2 in span_starts:
                    rels[(x1, x2)] = {}
            relations = example["relations"]
            for relation in relations:
                # the 'head' and 'child' annotations refer to the end token in the span
                # but we want the first token
                start = span_end_to_start[relation["head"]]
                end = span_end_to_start[relation["child"]]
                label = relation["label"]
                if label not in rels[(start, end)]:
                    rels[(start, end)][label] = 1.0
                    pos += 1
            # The annotation is complete, so fill in zero's where the data is missing
            for x1 in span_starts:
                for x2 in span_starts:
                    for label in REL_LABELS:
                        if label not in rels[(x1, x2)]:
                            rels[(x1, x2)][label] = 0.0
                            neg += 1
            doc._.rel = rels

            # only keeping documents with at least 1 positive case
            if pos > 0:
                docs.append(doc)
                count_pos += pos
                count_all += pos + neg
                count_sample += 1

        except Exception as e:
            msg.fail(f"Parse error: {e}")

    db = DocBin(docs=docs, store_user_data=True)
    db.to_disk(output_path)
```

And finally here is my pipeline config where I initialize the NER component with the vectors of another pretrained NER model:
```
[paths]
train = null
dev = null
vectors = null
init_tok2vec = null

[system]
gpu_allocator = "pytorch"
seed = 0

[nlp]
lang = "fr"
pipeline = ["transformer","ner","relation_extractor"]
batch_size = 128
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}

[components]

[components.transformer]
factory = "transformer"
max_batch_items = 4096
set_extra_annotations = {"@annotation_setters":"spacy-transformers.null_annotation_setter.v1"}

[components.transformer.model]
@architectures = "spacy-transformers.TransformerModel.v1"
name = "camembert-base"

[components.transformer.model.get_spans]
@span_getters = "spacy-transformers.strided_spans.v1"
window = 128
stride = 96

[components.transformer.model.tokenizer_config]
use_fast = true

[components.ner]
factory = "ner"
moves = null
update_with_oracle_cut_size = 100

[components.ner.model]
@architectures = "spacy.TransitionBasedParser.v2"
state_type = "ner"
extra_state_tokens = false
hidden_width = 64
maxout_pieces = 2
use_upper = false
nO = null

[components.ner.model.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0
pooling = {"@layers":"reduce_mean.v1"}
upstream = "*"

[components.relation_extractor]
factory = "relation_extractor"
threshold = 0.5

[components.relation_extractor.model]
@architectures = "rel_model.v1"

[components.relation_extractor.model.create_instance_tensor]
@architectures = "rel_instance_tensor.v1"

[components.relation_extractor.model.create_instance_tensor.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0

[components.relation_extractor.model.create_instance_tensor.tok2vec.pooling]
@layers = "reduce_mean.v1"

[components.relation_extractor.model.create_instance_tensor.pooling]
@layers = "reduce_mean.v1"

[components.relation_extractor.model.create_instance_tensor.get_instances]
@misc = "rel_instance_generator.v1"
max_length = 100

[components.relation_extractor.model.classification_layer]
@architectures = "rel_classification_layer.v1"
nI = null
nO = null

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 500
gold_preproc = false
limit = 0
augmenter = null

[training]
accumulate_gradient = 3
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
patience = 1600
max_epochs = 0
max_steps = 20000
eval_frequency = 200
frozen_components = []
before_to_disk = null

[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 2000
buffer = 256
get_length = null

[training.logger]
@loggers = "spacy.ConsoleLogger.v1"
progress_bar = true

[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001

[training.optimizer.learn_rate]
@schedules = "warmup_linear.v1"
warmup_steps = 250
total_steps = 20000
initial_rate = 0.00005

[training.score_weights]
rel_micro_p = 0.0
rel_micro_r = 0.0
rel_micro_f = 1.0

[pretraining]

[initialize]
vectors = "../../models/fr_ner_gpu"
init_tok2vec = ${paths.init_tok2vec}
vocab_data = null
lookups = null
before_init = null
after_init = null

[initialize.components]

[initialize.tokenizer]
```

I hope it's not too messy, thank you in advance for any help you can provide.

**Comment by svlandeg** [comment, marked as answer]
This pipeline component currently doesn't support training on NER and REL at the same time - instead it assumes that gold entities are already present when training the REL component. 

The underlying reason is that in spaCy 3.0, components are updated in isolation during training, which means that they don't see the predictions of any earlier components in the pipeline.

In the upcoming 3.1, this mechanism will become more transparent & extensible, and you'll be able to specify components that should set annotations on the predicted docs during training, so that downstream components can use those predictions as features or to build on them. In that case, you'd be able to use the predicted entities to train the REL.

For now, my advice would be to have a two-step training procedure: first create & train an NER that identifies all candidate mentions for your relations, then in a next step train a REL component that decides whether a link is correct or not. For the latter, you'll want to create a custom reader that already sets the `example.predicted.ents` as we do in the REL project.

**Reply by MightyGoldenOctopus** [reply]
@svlandeg So with the release of v3.1 I just have to specify `annotating_components = ["ner"]` in my pipeline config ?

**Reply by svlandeg** [reply]
Yes, I would hope that'd work! I haven't tested it yet though. Let us know how it goes!
